src/Pos/Context/Context.hs

```diff
@@ -0,0 +1,118 @@
+----------------------------------------------------------------------------
+-- NodeContext
+----------------------------------------------------------------------------
+
+-- | Data used for LRC syncronization. First value is __False__ iff
+-- LRC is running now. Second value is last epoch for which we have
+-- already computed LRC.
+type LrcSyncData = (Bool, EpochIndex)
+
+-- | NodeContext contains runtime context of node.
+data NodeContext ssc = NodeContext
+    { ncJLFile              :: !(Maybe (MVar FilePath))
+    -- @georgeee please add documentation when you see this comment
+    , ncSscContext          :: !(SscNodeContext ssc)
+    -- @georgeee please add documentation when you see this comment
+    , ncBlkSemaphore        :: !(MVar HeaderHash)
+    -- ^ Semaphore which manages access to block application.
+    -- Stored hash is a hash of last applied block.
+    , ncLrcSync             :: !(STM.TVar LrcSyncData)
+    -- ^ Primitive for synchronization with LRC.
+    , ncUserSecret          :: !(STM.TVar UserSecret)
+    -- ^ Secret keys (and path to file) which are used to send transactions
+    , ncBlockRetrievalQueue :: !(TBQueue (NodeId, NewestFirst NE (BlockHeader ssc)))
+    -- ^ Concurrent queue that holds block headers that are to be
+    -- downloaded.
+    , ncRecoveryHeader      :: !(STM.TMVar (NodeId, BlockHeader ssc))
+    -- ^ In case of recovery mode this variable holds the latest
+    -- header hash we know about so we can do chained block
+    -- requests. Invariant: this mvar is full iff we're more than
+    -- 'recoveryHeadersMessage' blocks deep relatively to some valid
+    -- header and we're downloading blocks. Every time we get block
+    -- that's more difficult than this one, we overwrite. Every time
+    -- we process some blocks and fail or see that we've downloaded
+    -- this header, we clean mvar.
+    , ncProgressHeader      :: !(STM.TMVar (BlockHeader ssc))
+    -- ^ Header of the last block that was downloaded in retrieving
+    -- queue. Is needed to show smooth prorgess on the frontend.
+    , ncUpdateSemaphore     :: !(MVar ConfirmedProposalState)
+    -- ^ A semaphore which is unlocked when update data is downloaded
+    -- and ready to apply.
+    , ncInvPropagationQueue :: !RelayInvQueue
+    -- ^ Queue is used in Relay framework,
+    -- it stores inv messages for earlier received data.
+    , ncLoggerConfig        :: !LoggerConfig
+    -- ^ Logger config, as taken/read from CLI.
+    , ncNodeParams          :: !NodeParams
+    -- ^ Params node is launched with
+    , ncShutdownFlag        :: !(STM.TVar Bool)
+    -- ^ If this flag is `True`, then workers should stop.
+    , ncShutdownNotifyQueue :: !(TBQueue ())
+    -- ^ A queue which is used to count how many workers have successfully
+    -- terminated.
+    , ncSendLock            :: !(Maybe (MVar ()))
+    -- ^ Exclusive lock for sending messages to other nodes
+    -- (if Nothing, no lock used).
+    , ncStartTime           :: !UTCTime
```

Comment:
:spiral_notepad: Must check that `UTCTime` is not used as if it's monotonic: what if the system time is adjusted by 500ms and a compare on two times is done?